vendor/github.com/danwinship/knftables/nftables.go

```diff
@@ -0,0 +1,444 @@
+/*
+Copyright 2023 Red Hat, Inc.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package knftables
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"os/exec"
+	"regexp"
+)
+
+// Interface is an interface for running nftables commands against a given family and table.
+type Interface interface {
```

Comment:
I've soured on this pattern of defining an interface and then returning it.  The Go pattern of "accept interfaces, return types" has started to make more sense to me.

The idea is that the receiver of an argument knows which methods they need, and can/should define a local interface with just those methods.  Trying to encapsulate that here just makes your library more complicated.

Not sure how well it applies here, but "type Interface interface" has become a code-smell to me.

Reply:
Yeah, I'm familiar with "accept interfaces, return types".

The problem here is `knftables.Fake`: we want it to support the same API as the real `knftables.Interface`, and we want consumers to be able to use it as such.

In a very quick search, I couldn't find any "accept interfaces, return types" articles that considered the situation where the API includes a fairly complex mock of itself (which makes heavy use of internal APIs) to help consumers with their own tests.